Register db

TijoloAzul opened the pull request:
Implementa o **registo**: 
1. O Login Client pede para se registar com um username e uma password
2. O Servidor verfifica se o nome já existe
2.a. Se existir responde com insucesso para o cliente
3. O Servidor guarda o novo registo na BD
4. É adquirido um novo login no nome do novo user
5. É enviado ao cliente uma mensagem de sucesso com o token de sessão
6. O cliente guarda o token indicando que já está logado

Implementa o **login** com BD
1. O Login Client pede ao servidor para abrir uma sessão, indicando o nome e a password
2. O servido verifica se o nome existe
2.a. Se não existir responde com insucesso para o cliente
3. O servido verifica se a password é correta
3.a. Se não fôr, responde com insucesso para o cliente
4. É adquirido um novo login no nome do novo user
5. É enviado ao cliente uma mensagem de sucesso com o token de sessão
6. O cliente guarda o token indicando que já está logado

TijoloAzul committed:
adding the capicity of logging in and register from client to database

TijoloAzul committed:
cleaning DB access

diogodinisf commented on src/main/java/pt/dinis/data/access/User.java:
```diff
@@ -30,60 +20,30 @@ public static void setNewUser(String username, String password, Connection conne
-        throw new NotFoundException();
-    }
-
-    public static String getName(int id, Connection connection) throws SQLException, NotFoundException {
```
Quando fores mostrar a um jogador que outros jogadores estão online vais precisar de mostrar o nome. Isto não será necessário?

TijoloAzul replied:
quando for preciso faz-se

diogodinisf replied:
refaz-se :P

diogodinisf reviewed (commented):
não sei se removeste cenas a mais

diogodinisf commented on src/main/java/pt/dinis/temporary/UserWorkerThread.java:
```diff
@@ -45,10 +47,32 @@ protected boolean working(Connection connection) {
+            return Dealer.sendMessage(Collections.singleton(id),
+                    new RegisterAnswer(UserMessage.AnswerType.ERROR, null, "Username already taken"));
+        }
+        User.createUser(message.getName(), message.getPassword(), connection);
```
não digo informações na consola. mas acho que vale muito a pena pelo meno avisar no logger que um utilizador acabou de se registar

TijoloAzul replied:
na boa... queres acrescentar, ou acrescento eu?

- [x] Acrescentar log

diogodinisf replied:
estás à vontade

diogodinisf commented on src/main/java/pt/dinis/temporary/UserWorkerThread.java:
```diff
@@ -45,10 +47,32 @@ protected boolean working(Connection connection) {
+        return Dealer.sendMessage(Collections.singleton(id),
+                new RegisterAnswer(UserMessage.AnswerType.SUCCESS, token, null));
+    }
+
```
aqui estás a usar uma abordagem diferente do caso do login. Não é que seja obrigatório fazer-se sempre igual. Mas convém ter uma razão para tal. Se houve diz :P

TijoloAzul replied:
diferente em quê?

diogodinisf replied:
para o login chamas uma rotina do Dealer que cria um token e envia a mensagem. Aqui crias o token e chamas a rotina de enviar mensagens do Dealer.

TijoloAzul replied:
Tens razão
- [x] Dealer é que faz a ligação entre o register no WorkerThread e o LoginManager.

diogodinisf commented on src/main/java/pt/dinis/temporary/UserWorkerThread.java:
```diff
@@ -45,10 +47,32 @@ protected boolean working(Connection connection) {
+                        new LoginAnswer(UserMessage.AnswerType.ERROR, null, "Wrong password"));
+            }
+            Display.info("Log in client " + id);
+            return Dealer.loginClient(id);
```
esta mensagem não seria para colocar aqui. De qualquer das formas. Correndo o risco de já termos falado sobre isto e de eu me ter esquecido. Eu faço login com um gajo, depois faço login com outro gajo e ele mete por cima do anterior. O anterior como vai ter um token desatualizado não vai cosneguir trabalhar mais.

TijoloAzul replied:
Explica melhor o teu problema...

diogodinisf replied:
o cliente A faz o login.
DE seguda o cliente B faz o login com o mesmo nome e o servidor deixa isso acontecer.

TijoloAzul replied:
Mas se B tem o mesmo nome... então B é o cliente A, certo?

TijoloAzul replied:
Isso ainda não está definido... por enquanto, o login na BD é apenas uma formalidade.
Nada a fazer.

diogodinisf reviewed (commented):
tinha-me esquecido de uma cenas

TijoloAzul committed:
pedido de registo no servidor foi transformado para:
1 - O pedido chega ao WorkerThread
2 - O WorkerThread pede aos serviços próximos da BD para registar o novo gajo
3 - O WorkerThread comunica ao Dealer para ele comunicar de volta com o cliente
4 - O Dealer pede um token ao LoginManager e responde ao cliente
